lapce-data/src/config.rs

```diff
@@ -538,30 +538,91 @@ impl Config {
-            .build()
-            .unwrap();
-        text_layout.size().width
+    /// Calculate the width of the character "W" (being the widest character)
+    /// in the editor's current font family and current font size.
+    pub fn editor_char_width(&self, text: &mut PietText) -> f64 {
+        Self::editor_text_size_internal(
+            self.editor.font_family(),
+            self.editor.font_size as f64,
+            text,
+            "W",
+        )
+        .width
+    }
+
+    /// Calculate the width of `text_to_measure` in the editor's current font family and font size.
+    pub fn editor_text_width(
+        &self,
+        text: &mut PietText,
+        text_to_measure: &str,
+    ) -> f64 {
+        Self::editor_text_size_internal(
+            self.editor.font_family(),
+            self.editor.font_size as f64,
+            text,
+            text_to_measure,
+        )
+        .width
+    }
+
+    /// Calculate the size of `text_to_measure` in the editor's current font family and font size.
+    pub fn editor_text_size(
+        &self,
+        text: &mut PietText,
+        text_to_measure: &str,
+    ) -> Size {
+        Self::editor_text_size_internal(
+            self.editor.font_family(),
+            self.editor.font_size as f64,
+            text,
+            text_to_measure,
+        )
     }
 
-    pub fn editor_text_size(&self, text: &mut PietText, c: &str) -> Size {
+    /// Efficiently calculate the size of a piece of text, without allocating.
+    /// This function should not be made public, use one of the public wrapper functions instead.
+    fn editor_text_size_internal(
+        font_family: FontFamily,
+        font_size: f64,
+        text: &mut PietText,
+        text_to_measure: &str,
+    ) -> Size {
+        // Lie about the lifetime of `text_to_measure`.
+        //
+        // The method `new_text_layout` will take ownership of its parameter
+        // and hold it inside the `text_layout` object. It will normally only do this efficiently
+        // for `&'static str`, and not other `&'a str`. It can safely do this for static strings
+        // because they are known to live for the lifetime of the program.
+        //
+        // `new_text_layout` can also work by taking ownership of an owned type such
+        // as String, Rc, or Arc. But they all require allocation. We want to measure
+        // `strs` with arbitrary lifetimes. If we 'cheat' by extending the lifetime of the
+        // `text_to_measure` (in this function only) then we can safely call `new_text_layout`
+        // because the `text_layout` value that is produced is local to this function and hence
+        // always dropped inside this function, and hence its lifetime is always stricly less
+        // than the lifetime of `text_to_measure`, irrespective of whether `text_to_measure`
+        // is actually static or not.
+        let static_str: &'static str =
+            unsafe { std::mem::transmute(text_to_measure) };
```

Comment:
We are also relying on implementation details here. Namely, that `new_text_layout` doesn't store our not-really-static string reference in a static/global/shared state. Which technically might change at any time in the future, so this transmute is really **very** unsafe and it would be a lot better if `new_text_layout` could just take any `&str` instead.